Add doc/example for ref assemblies, PDB and resources (#40735)
dotnet/docs: docs/fundamentals/runtime-libraries/system-reflection-emit-persistedassemblybuilder.md

ms.date: 05/10/2024
---
# Persisted dynamic assemblies in .NET

[!INCLUDE [context](includes/context.md)]

The <xref:System.Reflection.Emit.AssemblyBuilder.Save%2A?displayProperty=nameWithType> API wasn't originally ported to .NET (Core) because the implementation depended heavily on Windows-specific native code that also wasn't ported. In .NET 9 we added a fully managed `Reflection.Emit` implementation that supports saving. This implementation has no dependency on the pre-existing, runtime-specific `Reflection.Emit` implementation. That is, now there are two different implementations in .NET, runnable and persisted. To run the persisted assembly, first save it into a memory stream or a file, then load it back. Before `PersistedAssemblyBuilder`, because you could only run a generated assembly and not save it, it was difficult to debug these in-memory assemblies. Advantages of saving a dynamic assembly to a file are:

- You can verify the generated assembly with tools such as ILVerify, or decompile and manually examine it with tools such as ILSpy.
- The saved assembly can be loaded directly, no need to compile again, which can decrease application startup time.

> To create a `PersistedAssemblyBuilder` instance, use the `public PersistedAssemblyBuilder(AssemblyName name, Assembly coreAssembly, IEnumerable<CustomAttributeBuilder>? assemblyAttributes = null)` constructor. The `coreAssembly` parameter is used to resolve base runtime types and can be used for resolving reference assembly versioning:

Reviewer: ```suggestion
To create a `PersistedAssemblyBuilder` instance, use the <xref:System.Reflection.Emit.PersistedAssemblyBuilder.%23ctor(System.Reflection.AssemblyName,System.Reflection.Assembly,System.Collections.Generic.IEnumerable{System.Reflection.Emit.CustomAttributeBuilder})> constructor. The `coreAssembly` parameter is used to resolve base runtime types and can be used for resolving reference assembly versioning:
```

  :::code language="csharp" source="./snippets/System.Reflection.Emit/PersistedAssemblyBuilder/Overview/csharp/CreateAndRunAssembly.cs" id="Snippet2":::

## Set entry point for an executable

To set the entry point for an executable or to set other options for the assembly file, you can call the `public MetadataBuilder GenerateMetadata(out BlobBuilder ilStream, out BlobBuilder mappedFieldData)` method and use the populated metadata to generate the assembly with desired options, for example:

:::code language="csharp" source="./snippets/System.Reflection.Emit/PersistedAssemblyBuilder/Overview/csharp/GenerateMetadataSnippets.cs" id="Snippet1":::

## Emit symbols and generate PDB

> The symbols metadata is populated into the `pdbBuilder` out parameter when you call the `GenerateMetadata(out BlobBuilder ilStream, out BlobBuilder mappedFieldData, out MetadataBuilder pdbBuilder)` method on a `PersistedAssemblyBuilder` instance. To create an assembly with a portable PDB:

Reviewer: ```suggestion
The symbols metadata is populated into the `pdbBuilder` out parameter when you call the <xref:System.Reflection.Emit.PersistedAssemblyBuilder.GenerateMetadata(System.Reflection.Metadata.BlobBuilder@,System.Reflection.Metadata.BlobBuilder@)> method on a `PersistedAssemblyBuilder` instance. To create an assembly with a portable PDB:
```

> 1. Create `ISymbolDocumentWriter` instances with the `ModuleBuilder.DefineDocument(...)` method. While emitting the method's IL, also emit the corresponding symbol info.

Reviewer: ```suggestion
1. Create <xref:System.Diagnostics.SymbolStore.ISymbolDocumentWriter> instances with the <xref:System.Reflection.Emit.ModuleBuilder.DefineDocument(System.String,System.Guid,System.Guid,System.Guid)?displayProperty=nameWithType> method. While emitting the method's IL, also emit the corresponding symbol info.
```

> 2. Create a `PortablePdbBuilder` instance using the `pdbBuilder` instance produced by the `GenerateMetadata(...)` method.

Reviewer: ```suggestion
2. Create a <xref:System.Reflection.Metadata.Ecma335.PortablePdbBuilder> instance using the `pdbBuilder` instance produced by the <xref:System.Reflection.Emit.PersistedAssemblyBuilder.GenerateMetadata(System.Reflection.Metadata.BlobBuilder@,System.Reflection.Metadata.BlobBuilder@)> method.
```

> 3. Serialize the `PortablePdbBuilder` into a `Blob`, and write the `Blob` into a PDB file stream (only if you're generating a standalone PDB).

Reviewer: ```suggestion
3. Serialize the `PortablePdbBuilder` into a <xref:System.Reflection.Metadata.Blob>, and write the `Blob` into a PDB file stream (only if you're generating a standalone PDB).
```

> 4. Create a `DebugDirectoryBuilder` instance and add a `CodeViewEntry` (standalone PDB) or `EmbeddedPortablePdbEntry`.

Reviewer: ```suggestion
4. Create a <xref:System.Reflection.PortableExecutable.DebugDirectoryBuilder> instance and add a `CodeViewEntry` (standalone PDB) or `EmbeddedPortablePdbEntry`.
```

Reviewer: Where are CodeViewEntry and EmbeddedPortablePdbEntry defined?

Author: In DebugDirectoryBuilder, I added the link into your comment https://learn.microsoft.com/dotnet/api/system.reflection.portableexecutable.debugdirectorybuilder.addcodeviewentry

> 5. Set the optional `debugDirectoryBuilder` argument when creating the `PEBuilder` instance.

Reviewer: ```suggestion
5. Set the optional `debugDirectoryBuilder` argument when creating the <xref:System.Reflection.PortableExecutable.PEBuilder> instance.
```

    BlobBuilder peBlob = new BlobBuilder();
    peBuilder.Serialize(peBlob);

    using var fileStream = new FileStream("MyAssembly.exe", FileMode.Create, FileAccess.Write);
    peBlob.WriteContentTo(fileStream);
}

static DebugDirectoryBuilder GeneratePdb(MetadataBuilder pdbBuilder, ImmutableArray<int> rowCounts, MethodDefinitionHandle entryPointHandle)
{
    BlobBuilder portablePdbBlob = new BlobBuilder();
    PortablePdbBuilder portablePdbBuilder = new PortablePdbBuilder(pdbBuilder, rowCounts, entryPointHandle);
    BlobContentId pdbContentId = portablePdbBuilder.Serialize(portablePdbBlob);
    // In case saving PDB to a file
    using FileStream fileStream = new FileStream("MyAssemblyEmbeddedSource.pdb", FileMode.Create, FileAccess.Write);
    portablePdbBlob.WriteContentTo(fileStream);

    DebugDirectoryBuilder debugDirectoryBuilder = new DebugDirectoryBuilder();
    debugDirectoryBuilder.AddCodeViewEntry("MyAssemblyEmbeddedSource.pdb", pdbContentId, portablePdbBuilder.FormatVersion);
    // In case embedded in PE:
    // debugDirectoryBuilder.AddEmbeddedPortablePdbEntry(portablePdbBlob, portablePdbBuilder.FormatVersion);
    return debugDirectoryBuilder;
}
```

> Further you could add `CustomDebugInformation` by calling the `AddCustomDebugInformation` method from the `pdbBuilder` instance to add source embedding and source indexing etc. advanced PDB info.

Reviewer: ```suggestion
Further, you can add <xref:System.Reflection.Metadata.CustomDebugInformation> by calling the <xref:System.Reflection.Metadata.Ecma335.MetadataBuilder.AddCustomDebugInformation(System.Reflection.Metadata.EntityHandle,System.Reflection.Metadata.GuidHandle,System.Reflection.Metadata.BlobHandle)?displayProperty=nameWithType> method from the `pdbBuilder` instance to add source embedding and source indexing advanced PDB information.
```

```csharp
private static void EmbedSource(MetadataBuilder pdbBuilder)
{
    byte[] sourceBytes = File.ReadAllBytes("MySourceFile2.cs");
    BlobBuilder sourceBlob = new BlobBuilder();
    sourceBlob.WriteBytes(sourceBytes);
    pdbBuilder.AddCustomDebugInformation(MetadataTokens.DocumentHandle(1),
        pdbBuilder.GetOrAddGuid(new Guid("0E8A571B-6926-466E-B4AD-8AB04611F5FE")), pdbBuilder.GetOrAddBlob(sourceBlob));
}
```

## Add resources with PersistedAssemblyBuilder

> You can call `MetadataBuilder.AddManifestResource(...)` to add as many resources as needed. Streams must be concatenated into one `BlobBuilder` that you pass into the `ManagedPEBuilder` argument.

Reviewer: ```suggestion
You can call <xref:System.Reflection.Metadata.Ecma335.MetadataBuilder.AddManifestResource(System.Reflection.ManifestResourceAttributes,System.Reflection.Metadata.StringHandle,System.Reflection.Metadata.EntityHandle,System.UInt32)?displayProperty=nameWithType> to add as many resources as needed. Streams must be concatenated into one <xref:System.Reflection.Metadata.BlobBuilder> that you pass into the <xref:System.Reflection.PortableExecutable.ManagedPEBuilder> argument.
```

## Emit symbols and generate PDB

The symbols metadata is populated into the `pdbBuilder` out parameter when you call the <xref:System.Reflection.Emit.PersistedAssemblyBuilder.GenerateMetadata(System.Reflection.Metadata.BlobBuilder@,System.Reflection.Metadata.BlobBuilder@)> method on a `PersistedAssemblyBuilder` instance. To create an assembly with a portable PDB:

1. Create <xref:System.Diagnostics.SymbolStore.ISymbolDocumentWriter> instances with the <xref:System.Reflection.Emit.ModuleBuilder.DefineDocument(System.String,System.Guid,System.Guid,System.Guid)?displayProperty=nameWithType> method. While emitting the method's IL, also emit the corresponding symbol info.
2. Create a <xref:System.Reflection.Metadata.Ecma335.PortablePdbBuilder> instance using the `pdbBuilder` instance produced by the <xref:System.Reflection.Emit.PersistedAssemblyBuilder.GenerateMetadata(System.Reflection.Metadata.BlobBuilder@,System.Reflection.Metadata.BlobBuilder@)> method.
3. Serialize the `PortablePdbBuilder` into a <xref:System.Reflection.Metadata.Blob>, and write the `Blob` into a PDB file stream (only if you're generating a standalone PDB).

> 4. Create a <xref:System.Reflection.PortableExecutable.DebugDirectoryBuilder> instance and add a <xref:System.Reflection.PortableExecutable.DebugDirectoryBuilder.CodeViewEntry%2A?displayProperty=nameWithType> (standalone PDB) or <xref:System.Reflection.PortableExecutable.DebugDirectoryBuilder.EmbeddedPortablePdbEntry%2A?displayProperty=nameWithType>. 

Author: ```suggestion
4. Create a <xref:System.Reflection.PortableExecutable.DebugDirectoryBuilder> instance and add a <xref:System.Reflection.PortableExecutable.DebugDirectoryBuilder.CodeViewEntry%2A?displayProperty=nameWithType> (standalone PDB) or <xref:System.Reflection.PortableExecutable.DebugDirectoryBuilder.EmbeddedPortablePdbEntry%2A?displayProperty=nameWithType>.
```
